src/intercept_desc.c

```diff
@@ -0,0 +1,632 @@
+
+		/* a function entry point in .text, mark it */
+		mark_jump(desc, address);
+
+		if (has_no_syscall(address, syms[i].st_size))
+			mark_skip_range(desc, address, syms[i].st_size);
+	}
+}
+
+/*
+ * has_pow2_count
+ * Checks if the number of patches in a struct intercept_desc
+ * is a power of two or not.
+ */
+static bool
+has_pow2_count(const struct intercept_desc *desc)
+{
+	return (desc->count & (desc->count - 1)) == 0;
+}
+
+/*
+ * add_new_patch
+ * Acquires a new patch entry, and allocates memory for it if
+ * needed.
+ */
+static struct patch_desc *
+add_new_patch(struct intercept_desc *desc)
+{
+	if (desc->count == 0) {
+
+		/* initial allocation */
+		desc->items = xmmap_anon(sizeof(desc->items[0]));
+
+	} else if (has_pow2_count(desc)) {
+
+		/* if count is a power of two, double the allocate space */
+		size_t size = desc->count * sizeof(desc->items[0]);
+
+		desc->items = xmremap(desc->items, size, 2 * size);
+	}
+
+	return &(desc->items[desc->count++]);
+}
+
+/*
+ * crawl_text
+ * Crawl the text section, disassembling it all.
+ * This routine collects information about potential addresses to patch.
+ *
+ * The addresses of all syscall instructions are stored, together with
+ * a description of the preceding, and following instructions.
+ *
+ * A lookup table of all addresses which appear as jump destination is
+ * generated, to help determine later, whether an instruction is suitable
+ * for being overwritten -- of course, if an instruction is a jump destination,
+ * it can not be merged with the preceding instruction to create a
+ * new larger one.
+ *
+ * Note: The actual patching can not yet be done in this disassembling phase,
+ * as it is not known in advance, which addresses are jump destinations.
+ */
+static void
+crawl_text(struct intercept_desc *desc)
+{
+	unsigned char *code = desc->text_start;
+
+	/*
+	 * Remember the previous three instructions, while
+	 * disassembling the code instruction by instruction in the
+	 * while loop below.
+	 */
+	struct intercept_disasm_result prevs[3] = {{0, }};
+
+	struct range *skip = desc->skip_ranges;
+
+	/*
+	 * How many previous instructions were decoded before this one,
+	 * and stored in the prevs array. Usually three, except for the
+	 * beginning of the text section -- the first instruction naturally
+	 * has no previous instruction.
+	 */
+	unsigned has_prevs = 0;
+	struct intercept_disasm_context *context =
+	    intercept_disasm_init(desc->text_start, desc->text_end);
+
+	while (code <= desc->text_end) {
+
+
+		while (skip->address != NULL && code > skip->address)
+			++skip;
+
+		if (code == skip->address) {
+			code += skip->size;
+			continue;
+		}
+
+		struct intercept_disasm_result result;
+
+		result = intercept_disasm_next_instruction(context, code);
+
+		if (result.length == 0) {
+			++code;
+			continue;
+		}
+
+		if (result.has_ip_relative_opr)
+			mark_jump(desc, result.rip_ref_addr);
+
+		if (result.is_nop && result.length >= 7)
```

Comment:
It would be cleaner to create #define for any integer value you are comparing. In the above code using the comparison >=7 isn't obvious as to why it's this value.